### PR TITLE
[Prim] Implement group_norm_backward

### PR DESCRIPTION
**Pull request by malfet**
Test plan: CI, i.e. `python3 test_decomp.py -v -k test_comprehensive_nn_functional_group_norm` plus:
```
#!/usr/bin/env python3.8
import torch

func = torch.ops.aten.native_group_norm_backward.default
decomp =  torch._decomp.decomposition_table[func]
for args in (
        (torch.rand(1, 6, 3), torch.rand(1, 6, 3), torch.rand(1, 2), torch.rand(1, 2), torch.rand(6), 1, 6, 3, 2, [True, True, True]),
        (torch.rand(64, 768, 7, 7), torch.rand(64, 768, 7, 7), torch.rand(64, 1), torch.rand(64, 1), torch.rand(768), 64, 768, 49, 1, [True, True, True])):
    nrc=func(*args)
    drc=decomp(*args)
    for i in range(len(nrc)):
       print(i, torch.max(nrc[i]-drc[i]))
    print(all(torch.allclose(x, y) for (x, y) in zip(nrc, drc)))
```

@diff-train-skip-merge

### CODE REVIEW

**Comment by Chillee** [comment]
I think you actually do fail CUDA tests for float16/bfloat16 accuracy :) The lack of tests for masks other than `[True, False, False]` is ... unfortunate though.

**Comment by malfet** [comment]
> I think you actually do fail CUDA tests for float16/bfloat16 accuracy :) The lack of tests for masks other than `[True, False, False]` is ... unfortunate though.

Good point, haven't realized that float16 tests on CPU are no-ops (say should have been reported as skipped then, or may be not even generated)

Current iteration should be clean on CUDA, will  extend the test to cover other masks

File: torch/_decomp/decompositions.py
Comment (ngimel) [inline torch/_decomp/decompositions.py:870]: probably do @pw_cast_for_opmath instead of this, to make sure all inputs are converted

File: torch/_decomp/decompositions.py
Comment (ngimel) [inline torch/_decomp/decompositions.py:877]: should there be a check that it evenly divides?

File: torch/_decomp/decompositions.py
Comment (ngimel) [inline torch/_decomp/decompositions.py:881]: `gamma.unsqueeze(0)`, `unsqueeze` is typically easier for backends to reason about

File: torch/_decomp/decompositions.py
Comment (ngimel) [inline torch/_decomp/decompositions.py:897]: unsqueeze

File: torch/_decomp/decompositions.py
Comment (ngimel) [inline torch/_decomp/decompositions.py:909]: but `db` has size (N,C)?

Comment (malfet) [reply torch/_decomp/decompositions.py:909]: You are right, it should be `db.sum(0)`

File: torch/_decomp/decompositions.py
Comment (Chillee) [reply torch/_decomp/decompositions.py]: wait is `pw_cast_for_opmath` the right decorator to use here? Do you want to upcast all inputs to fp32?

Comment (ngimel) [reply torch/_decomp/decompositions.py]: idk how it's implemented in eager, but morally yes, all ops should be in fp32? E.g. in batch norm backward all intermediate computations are in fp32. 
We probably should be able to check that all inputs are of expected type, and type promotion decorations prevent us from doing so, but oh well.

File: torch/_decomp/decompositions.py
Comment (malfet) [reply torch/_decomp/decompositions.py]: Updated, not sure if this is what you've meant though...

File: torch/_decomp/decompositions.py
Comment (Chillee) [inline torch/_decomp/decompositions.py:869]: I think @ngimel is suggesting to do something like https://github.com/pytorch/pytorch/blob/master/torch/_refs/__init__.py#L560

**Review by Chillee — APPROVED** [review]
LGTM, although I'd wait for Natalia's approval :P 

Would be nice to get the test in core somewhere.

Also, if you wanna try it, might be worth benchmarking how well Inductor lowers this op :)

https://github.com/pytorch/torchdynamo/pull/785

File: torch/_decomp/decompositions.py
Comment (Chillee) [inline torch/_decomp/decompositions.py:913]: Actually, I saw some shape errors with this line when patching this decomp in and trying out @eellison 's per-operator benchmarks.

(from torchdynamo)

```
python ./benchmarks/microbenchmarks/operatorbench.py --op=aten.native_group_norm_backward.default --dtype=float16 --suite=timm
```

Comment (ngimel) [reply torch/_decomp/decompositions.py:913]: Yeah similar to `d_bias` there should be `.sum(0)` here

Comment (malfet) [reply torch/_decomp/decompositions.py]: Added `.sum(0)`, though it still need reshape here.

**Review by ngimel — APPROVED** [review]
Thanks!

**Pull request closed without merging**
